impl/spanner/batch/batch_test.go

```diff
@@ -0,0 +1,160 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package batch
+
+import (
+	"context"
+	"testing"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/keytransparency/core/integration/storagetest"
+	"github.com/google/keytransparency/core/sequencer"
+	"github.com/google/keytransparency/impl/spanner/directory"
+	"github.com/google/keytransparency/impl/spanner/testutil"
+	"github.com/google/trillian/crypto/keyspb"
+
+	dtype "github.com/google/keytransparency/core/directory"
+	spb "github.com/google/keytransparency/core/sequencer/sequencer_go_proto"
+	ktspanner "github.com/google/keytransparency/impl/spanner"
+	tpb "github.com/google/trillian"
+)
+
+func NewForTest(ctx context.Context, t *testing.T, dirID string) (*Table, func()) {
+	t.Helper()
+	ddl, err := ktspanner.ReadDDL()
+	if err != nil {
+		t.Fatal(err)
+	}
+	client, cleanup := testutil.CreateDatabase(ctx, t, ddl)
+	b := New(client)
+
+	if err := directory.New(client).Write(ctx, &dtype.Directory{
+		DirectoryID: dirID,
+		Map:         &tpb.Tree{},
+		Log:         &tpb.Tree{},
+		VRFPriv:     &keyspb.PrivateKey{Der: []byte("privkeybytes")},
+	}); err != nil {
+		t.Fatalf("directories.Write(%v): %v", dirID, err)
+	}
+
+	return b, cleanup
+}
+
+func TestNewForTest(t *testing.T) {
+	ctx := context.Background()
+	directoryID := "new"
+	_, done := NewForTest(ctx, t, directoryID)
+	defer done()
+}
+
+func TestBatchIntegration(t *testing.T) {
+	storageFactory :=
+		func(ctx context.Context, t *testing.T, dirID string) (sequencer.Batcher, func(context.Context)) {
+			b, done := NewForTest(ctx, t, dirID)
+			return b, func(_ context.Context) { done() }
+		}
+
+	storagetest.RunBatchStorageTests(t, storageFactory)
+}
+
+func TestWriteBatch(t *testing.T) {
+	ctx := context.Background()
+	directoryID := "writebatch"
+	m, done := NewForTest(ctx, t, directoryID)
+	defer done()
+	for _, tc := range []struct {
+		rev     int64
+		wantErr bool
+		sources []*spb.MapMetadata_SourceSlice
+	}{
+		// Tests are cumulative.
+		{rev: 0, sources: []*spb.MapMetadata_SourceSlice{{LogId: 1, HighestExclusive: 10}}},
+		{rev: 0, sources: []*spb.MapMetadata_SourceSlice{{LogId: 1, HighestExclusive: 11}}, wantErr: true},
+		{rev: 0, sources: []*spb.MapMetadata_SourceSlice{{LogId: 2, HighestExclusive: 20}}, wantErr: true},
+		{rev: 0, sources: []*spb.MapMetadata_SourceSlice{}, wantErr: true},
+		{rev: 1, sources: []*spb.MapMetadata_SourceSlice{}},
+		{rev: 1, sources: []*spb.MapMetadata_SourceSlice{{LogId: 1, HighestExclusive: 10}}, wantErr: true},
+	} {
+		err := m.WriteBatchSources(ctx, directoryID, tc.rev, &spb.MapMetadata{Sources: tc.sources})
+		if got, want := err != nil, tc.wantErr; got != want {
+			t.Errorf("WriteBatchSources(%v, %v): err: %v. code: %v, want %v",
+				tc.rev, tc.sources, err, got, want)
+		}
+	}
+}
+
+func TestReadBatch(t *testing.T) {
```

Comment:
Add coverage for reading a revision which doesn't yet exist?

Reply:
done